backend/src/apiserver/resource/semaphore_manager.go

```diff
@@ -0,0 +1,106 @@
+package resource
+
+import (
+	"context"
+	"os"
+
+	"github.com/golang/glog"
+	"github.com/kubeflow/pipelines/backend/src/apiserver/client"
+	"github.com/kubeflow/pipelines/backend/src/common/util"
+	v1 "k8s.io/api/core/v1"
+	k8errors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+)
+
+// Common constants for semaphore functionality
+const (
+	SemaphoreConfigName = "semaphore-config"
+)
+
+// Common Kubernetes labels used across KFP resources
+const (
+	LabelApp       = "kubeflow-pipelines"
+	LabelManagedBy = "kubeflow-pipelines-apiserver"
+)
+
+const (
+	LabelKeyApp       = "app"
+	LabelKeyComponent = "component"
+	LabelKeyManagedBy = "app.kubernetes.io/managed-by"
+)
+
+func createKFPLabels(component string) map[string]string {
+	return map[string]string{
+		LabelKeyApp:       LabelApp,
+		LabelKeyComponent: component,
+		LabelKeyManagedBy: LabelManagedBy,
+	}
+}
+
+type SemaphoreManager struct {
+	k8sClient client.KubernetesCoreInterface
+}
+
+func NewSemaphoreManager(k8sClient client.KubernetesCoreInterface) *SemaphoreManager {
+	return &SemaphoreManager{k8sClient: k8sClient}
+}
+
+// EnsureSemaphoreConfigMap creates a ConfigMap entry for the given semaphore key with default limit of 1.
+// Only creates new entries - preserves existing user edits.
+func (s *SemaphoreManager) EnsureSemaphoreConfigMap(ctx context.Context, namespace, semaphoreKey string) error {
+	if semaphoreKey == "" {
+		return nil
+	}
+
+	configMapName := s.getSemaphoreConfigMapName()
+	configMapClient := s.k8sClient.ConfigMapClient(namespace)
+
+	// Get existing ConfigMap
+	configMap, err := configMapClient.Get(ctx, configMapName, metav1.GetOptions{})
+	if k8errors.IsNotFound(err) {
+		// Create new ConfigMap
+		configMap = &v1.ConfigMap{
+			ObjectMeta: metav1.ObjectMeta{
+				Name:      configMapName,
+				Namespace: namespace,
+				Labels:    createKFPLabels(SemaphoreConfigName),
+			},
+			Data: map[string]string{
+				semaphoreKey: "1", // Default limit of 1 (mutex behavior)
+			},
+		}
+		_, err = configMapClient.Create(ctx, configMap, metav1.CreateOptions{})
+		if err != nil {
```

Comment:
I think we should retry if the error is due to the object already existing to account for race conditions of two pipeline runs being submitted at the same time.